lib/neo4j/shared/declared_property_manager.rb

```diff
@@ -0,0 +1,94 @@
+module Neo4j::Shared
+  class DeclaredPropertyManager
+    include Neo4j::Shared::TypeConverters
+
+    attr_reader :klass
+
+    def initialize(klass)
+      @klass = klass
+    end
+
+    def register(property)
+      @_attributes_nil_hash = nil
+      registered_properties[property.name] = property
+      register_magic_typecaster(property) if property.magic_typecaster
+      declared_property_defaults[property.name] = property.default_value if property.default_value
+    end
+
+    def declared_property_defaults
+      @_default_property_values ||= {}
+    end
+
+    def registered_properties
+      @_registered_properties ||= {}
+    end
+
+    def attributes_nil_hash
+      @_attributes_nil_hash ||= {}.tap { |attr_hash| registered_properties.each_pair { |k, _v| attr_hash[k.to_s] = nil } }.freeze
```

Comment:
Maybe you're thinking of `Hash#keys`?  I think that the reason that `Hash#each_key` exists is because `Hash#keys` makes an additional `Array`.  In my quick tests, actually a bit surprisingly `hash.keys.each` is fastest followed by `hash.each_key` followed by `hash.each_pair`

```
2.2.0 :006 > Benchmark.realtime { 1000.times { h.each_key {|k| 1 } } }
 => 0.6483775430097012
2.2.0 :007 > Benchmark.realtime { 1000.times { h.each_key {|k| 1 } } }
 => 0.5919963799969992
2.2.0 :008 > Benchmark.realtime { 1000.times { h.each_key {|k| 1 } } }
 => 0.5759841660037637
2.2.0 :009 > Benchmark.realtime { 1000.times { h.each_key {|k| 1 } } }
 => 0.5980105880007613
2.2.0 :010 > Benchmark.realtime { 1000.times { h.each_pair {|k| 1 } } }
 => 0.9172219970059814
2.2.0 :011 > Benchmark.realtime { 1000.times { h.each_pair {|k| 1 } } }
 => 0.9443715289962711
2.2.0 :012 > Benchmark.realtime { 1000.times { h.each_pair {|k| 1 } } }
 => 0.9549578059959458
2.2.0 :013 > Benchmark.realtime { 1000.times { h.keys.each {|k| 1 } } }
 => 0.36982293298933655
2.2.0 :014 > Benchmark.realtime { 1000.times { h.keys.each {|k| 1 } } }
 => 0.35002660199825186
2.2.0 :015 > Benchmark.realtime { 1000.times { h.keys.each {|k| 1 } } }
 => 0.36031401000218466
```